// source-map-explorer/dist/html.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true,
});
exports.generateHtml = generateHtml;

var _btoa = _interopRequireDefault(require('btoa'));

var _ejs = _interopRequireDefault(require('ejs'));

var _fs = _interopRequireDefault(require('fs'));

var _path = _interopRequireDefault(require('path'));

var _escapeHtml = _interopRequireDefault(require('escape-html'));

var _helpers = require('./helpers');

function _interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : { default: obj };
}

/**
 * Generate HTML file content for specified files
 */
function generateHtml(exploreResults) {
  const assets = {
    webtreemapJs: (0, _btoa.default)(
      _fs.default.readFileSync(require.resolve('./vendor/webtreemap.js'))
    ),
    webtreemapCss: (0, _btoa.default)(
      _fs.default.readFileSync(require.resolve('./vendor/webtreemap.css'))
    ),
  }; // Create a combined bundle if applicable

  if (exploreResults.length > 1) {
    exploreResults = [makeMergedBundle(exploreResults)].concat(exploreResults);
  } // Get bundles info to generate select element

  const bundles = exploreResults.map(data => ({
    name: normalizeBundleName(data.bundleName),
    size: (0, _helpers.formatBytes)(data.totalBytes),
  })); // Get webtreemap data to update map on bundle select

  const treeDataMap = exploreResults.reduce((result, data) => {
    result[data.bundleName] = getWebTreeMapData(data.files);
    return result;
  }, {});
  const template = (0, _helpers.getFileContent)(_path.default.join(__dirname, 'tree-viz.ejs'));
  return _ejs.default.render(template, {
    bundles,
    treeDataMap,
    webtreemapJs: assets.webtreemapJs,
    webtreemapCss: assets.webtreemapCss,
  });
}
/**
 * Create a combined result where each of the inputs is a separate node under the root
 */

function makeMergedBundle(exploreResults) {
  let totalBytes = 0;
  const files = {}; // Remove any common prefix to keep the visualization as simple as possible.

  const commonPrefix = (0, _helpers.getCommonPathPrefix)(exploreResults.map(r => r.bundleName));

  for (const result of exploreResults) {
    totalBytes += result.totalBytes;
    const prefix = result.bundleName.slice(commonPrefix.length);
    Object.entries(result.files).forEach(([fileName, size]) => {
      files[`${prefix}/${fileName}`] = size;
    });
  }

  return {
    bundleName: '[combined]',
    totalBytes,
    unmappedBytes: 0,
    files,
  };
}

const ESCAPE_BACKSLASH_REGEX = /\\/g;
/** Replaces `\` by `\\` in Windows path so that string is escaped when inserted in template */

function normalizeBundleName(name) {
  return name.replace(ESCAPE_BACKSLASH_REGEX, '\\\\');
}

/**
 * Covert file size map to webtreemap data
 */
function getWebTreeMapData(files) {
  const treeData = newNode('/');

  for (const source in files) {
    addNode(source, files[source], treeData);
  }

  addSizeToTitle(treeData, treeData.data['$area']);
  return treeData;
}

function newNode(name) {
  return {
    name: (0, _escapeHtml.default)(name),
    data: {
      $area: 0,
    },
  };
}

function addNode(source, size, treeData) {
  const parts = source.split('/');
  let node = treeData;
  node.data['$area'] += size;
  parts.forEach(part => {
    if (!node.children) {
      node.children = [];
    }

    let child = node.children.find(child => child.name === part);

    if (!child) {
      child = newNode(part);
      node.children.push(child);
    }

    node = child;
    node.data['$area'] += size;
  });
}

function addSizeToTitle(node, total) {
  const size = node.data['$area'];
  node.name += ` • ${(0, _helpers.formatBytes)(size)} • ${(0, _helpers.formatPercent)(
    size,
    total,
    1
  )}%`;

  if (node.children) {
    node.children.forEach(child => {
      addSizeToTitle(child, total);
    });
  }
}
